bulk inserts

nhachicha opened the pull request:
Fixes https://github.com/realm/realm-java/issues/1684 

# Benchmarks with 100_000 objects
## Before (using `copyToRealm`)
![benchmark_batchwrite](https://cloud.githubusercontent.com/assets/1793238/16064422/fdd88640-3298-11e6-9601-b9bdfbaabab2.png)


## Using Bulk Inserts   
![benchmark_batchwrite](https://cloud.githubusercontent.com/assets/1793238/16064440/2754c560-3299-11e6-8439-d64f6779eb44.png)

# Cap 

The new API accepts a list of `Collection` instead of a list of `Iterator` because we need to know the size of the collection, in order to optimise the `Map` that acts as a cache for the previously inserted objects.
here is an Android _traceview_ dump that stress the importance of initialising the `IdentityMap` to avoid the expensive `rehash` invocation for each `put` operation.

## Without initialising the Map
<img width="780" src="https://cloud.githubusercontent.com/assets/1793238/16064630/cbf7d12e-329a-11e6-844f-1e95a2d9eeb5.png">
we spend 35% of the overall time in the `put` operation 
<img width="693" src="https://cloud.githubusercontent.com/assets/1793238/16064676/39c1bcf6-329b-11e6-915a-986eaeaaf0de.png">
the `rehash` could be avoided if we initialise the Map with the appropriate size

## Initialising the Map with the size of the Collection
<img width="628" alt="after_rehash_opts" src="https://cloud.githubusercontent.com/assets/1793238/16064663/19aa16de-329b-11e6-9341-ac05cec7d0f1.png">

Although we reduced the overall time spent in the Map, we can't reach the performance of lowlevel Realm, because the lowlevel Realm doesn't need the `Map` to cache objects.

nhachicha commented:
@realm/java the CI is failing until https://github.com/realm/realm-java/pull/2990/commits/ac859080c506170782bfeac65bcbd85d26ed0b3b is merged from release to master

stk1m1 commented:
retest this please

stk1m1 commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,491 @@
+        } catch (RealmException ignored) {
+        }
+
+        realm.commitTransaction();
```
Just a nit pick. Shouldn't this be 

```java
} finally {
    realm.cancelTransaction();
}
```
?

kneth commented on realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java:
```diff
@@ -656,6 +655,248 @@ public static AllTypes copy(Realm realm, AllTypes newObject, boolean update, Map
+            object = (AllTypes) objects.next();
+            long rowIndex = Table.nativeAddEmptyRow(tableNativePtr, 1);
+            cache.put(object, rowIndex);
+            String realmGet$columnString =  ((AllTypesRealmProxyInterface)object).realmGet$columnString();
```
Can't you share code with the `insert` above?

nhachicha replied:
you can but since this PR is about optimisation, I prefer inlining the code, rather than the cost of calling another method, it's generated code anyway

kneth commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+     *
+     * @param objects RealmObjects to insert.
+     */
+    public void insertToRealm(Collection<? extends RealmModel> objects) {
```
Will `insertIntoRealm` be a better name?

nhachicha replied:
I try to stay close to the existing `copyToRealm` & `copyToRealmOrUpdate` naming

cmelchior replied:
I see the connection but IMO `insertToRealm` sounds wrong somehow. Any of these would be better?
- `insert()`
- `insertObject()`
- `insertIntoRealm()` is probably more grammatically correct that `toRealm()`. Still sounds weird IMO.

kneth commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
     }
 
     /**
+     * Bulk insert of a list of RealmObjects, this is faster than {@link #copyToRealm(Iterable)} since it
```
"faster" -> "generally faster" (if a class only has one field, I don't think it will be faster)

kneth commented on realm/realm-library/src/main/java/io/realm/internal/LinkView.java:
```diff
@@ -167,15 +167,15 @@ private void checkImmutable() {
     }
 
-    static native void nativeClose(long nativeLinkViewPtr);
+    public static native void nativeClose(long nativeLinkViewPtr);
```
Can we get away with `protected` instead of `public`?

nhachicha replied:
I'm afraid not, the generated code is inside `io.realm` package

kneth commented:
After first pass: looks good. I'll do a second pass later.

stk1m1 commented on realm/realm-library/src/main/java/io/realm/internal/CheckedRow.java:
```diff
@@ -126,7 +126,7 @@ public void setNull(long columnIndex) {
     protected native int nativeGetMixedType(long nativePtr, long columnIndex);
     protected native Mixed nativeGetMixed(long nativeRowPtr, long columnIndex);
-    protected native long nativeGetLinkView(long nativePtr, long columnIndex);
+    public static native long nativeGetLinkView(long nativePtr, long columnIndex);
```
When you modify `naive*` method as a `static`, don't forget to change the signature of corresponding JNI function arguments. The argument should be `(JNIEnv, jclass,...)` than `(JNIEnv, jobject,...)` could cause timeout on some devices. Look this guy's [comment](https://github.com/realm/realm-java/pull/2995#issuecomment-226070423)

stk1m1 commented on realm/realm-library/src/main/java/io/realm/internal/LinkView.java:
```diff
@@ -167,15 +167,15 @@ private void checkImmutable() {
     native long nativeGetRow(long nativeLinkViewPtr, long pos);
     private native long nativeGetTargetRowIndex(long nativeLinkViewPtr, long pos);
-    private native void nativeAdd(long nativeLinkViewPtr, long rowIndex);
+    public static native void nativeAdd(long nativeLinkViewPtr, long rowIndex);
```
same

stk1m1 commented on realm/realm-library/src/main/java/io/realm/internal/LinkView.java:
```diff
@@ -167,15 +167,15 @@ private void checkImmutable() {
     private native void nativeMove(long nativeLinkViewPtr, long oldPos, long newPos);
     private native void nativeRemove(long nativeLinkViewPtr, long pos);
-    private native void nativeClear(long nativeLinkViewPtr);
+    public static native void nativeClear(long nativeLinkViewPtr);
```
same

stk1m1 commented on realm/realm-library/src/main/java/io/realm/internal/Table.java:
```diff
@@ -1548,7 +1552,7 @@ public static String tableNameToClassName(String tableName) {
     private native void nativeRemoveLast(long nativeTablePtr);
     private native void nativeMoveLastOver(long nativeTablePtr, long rowIndex);
-    private native long nativeAddEmptyRow(long nativeTablePtr, long rows);
+    public static native long nativeAddEmptyRow(long nativeTablePtr, long rows);
```
Same here. There are 12 more static methods below.

kneth replied:
Do they have to be `public`?

stk1m1 commented:
Looks awesome! I especially love `traceview` graph!

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    .emitStatement("Short %s =  ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetLong(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
+                    if(isUpdate) {
```
`if(` -> `if (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if ("java.lang.Byte".equals(fieldType)) {
+            writer
+                    .emitStatement("Byte %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
`=  (` -> `= (` (remove extra space)

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    .emitStatement("Byte %s =  ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetLong(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
+                    if(isUpdate) {
```
`if(` -> `if (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    .emitStatement("byte[] %s =  ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetByteArray(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
+                    if(isUpdate) {
```
`if(` -> `if (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    .emitStatement("java.util.Date %s =  ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetTimestamp(tableNativePtr, columnInfo.%sIndex, rowIndex, %s.getTime())", fieldName, getter);
+                    if(isUpdate) {
```
`if(`-> `if (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    .emitStatement("String %s =  ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetString(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
+                    if(isUpdate) {
```
`if(` -> `if (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                                Utils.getProxyClassSimpleName(field),
+                                fieldName)
+                        .endControlFlow()
+                        .emitStatement("Table.nativeSetLink(tableNativePtr, columnInfo.%sIndex, rowIndex, cache%s)", fieldName, fieldName)
```
Why don't you swap the last two arguments? :-)

cmelchior replied:
You can use named parameters and only have fieldName once

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                , columnInfoClassName(), columnInfoClassName(), className);
+        writer.emitStatement("%s object = null", className);
+
+        writer.beginControlFlow("while(objects.hasNext())");
```
`e(` -> `e (`

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        }
+        writer.emitStatement("%s object = null", className);
+
+        writer.beginControlFlow("while(objects.hasNext())");
```
`e(` -> `e (`

kneth commented:
It looks good to me. I like that `Table`'s native methods are called as it will give the best performance.

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    }
+                    writer.endControlFlow();
+
+        } else if ("java.lang.Integer".equals(fieldType)) {
```
Couldn't all these be collapsed into the `Long` case? I mean the java type is short/int already, so any overflow have already happen meaning it should be safe to case it to Long now?

nhachicha replied:
Somehow you can't cast the boxed type.
this fails in the IDE, ex:
```java
Short s = Short.valueOf((short)10);
Long l = (Long)s;
```
will fail with _Inconvertible types ..._

cmelchior replied:
Ah, yes, that makes sense, since Short and Long both extends from Number. You should be able to cast to Number though?

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                    }
+                    writer.endControlFlow();
+
+        } else if ("java.lang.Byte".equals(fieldType)) {
```
Byte could be cast to long as well?

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if ("java.lang.Float".equals(fieldType)) {
+            writer
+                    .emitStatement("Float %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
Whitespace after =

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        writer.beginMethod(
+                "long", // Return type
+                "insert", // Method name
+                EnumSet.of(Modifier.PUBLIC, Modifier.STATIC), // Modifiers
```
PROTECTED should be enough?

nhachicha replied:
nope, the interface (`internal` package) is called from `io.realm`

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .emitStatement("RealmList<%s> %sList = ((%s) object).%s()",
+                                Utils.getGenericType(field), fieldName, interfaceName, getter)
+                        .beginControlFlow("if (%sList != null)", fieldName)
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
Indent all these statements so they follow the generated flow
```
.beginControlFlow()
    .emitStatement()
.endControlFlow()
```

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        writer.beginControlFlow("while(objects.hasNext())");
+        writer.emitStatement("object = (%s) objects.next()", className);
+
+        writer.emitStatement("long rowIndex = Table.nativeAddEmptyRow(tableNativePtr, 1)");
```
Can we create all the empty rows we need in one go?

nhachicha replied:
Yes, but there are 2 problems with that:

1- You don't know before you traverse the graph how many new rows you need, ex on some cyclic dependency you're going to skip creating a new rows, also in `insertOrUpdate` you may end up updating the row rather than inserting a new one.

2- You need to keep track  of the next available row index  as you may skip some (point 1), at the end AFAIK there's no Core method to deallocate the unused rows, (you may end up iterating through the unused indexes & call delete?)

cmelchior replied:
Good point. Keeping it like it is seems better then 👍

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        writer.beginMethod(
+                "long", // Return type
+                "insertOrUpdate", // Method name
+                EnumSet.of(Modifier.PUBLIC, Modifier.STATIC), // Modifiers
```
PROTECTED?

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyMediatorGenerator.java:
```diff
@@ -240,6 +245,94 @@ public void emitStatement(int i, JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "void",
+                "insertToRealm",
```
InsertToRealm sounds wrong somehow. InsertInRealm? or just InsertObject?

nhachicha replied:
this is internal API anyway I'm fine with `insertObject`

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyMediatorGenerator.java:
```diff
@@ -240,6 +245,94 @@ public void emitStatement(int i, JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "void",
+                "insertOrUpdateToRealm",
```
InsertOrUpdateObject?

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java:
```diff
@@ -656,6 +655,248 @@ public static AllTypes copy(Realm realm, AllTypes newObject, boolean update, Map
+        AllTypes object = null;
+        while(objects.hasNext()) {
+            object = (AllTypes) objects.next();
+            long rowIndex = Table.nativeAddEmptyRow(tableNativePtr, 1);
```
Move this outside and allocate everything in one go?

nhachicha replied:
see my previous point

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -104,6 +105,14 @@
      */
     public abstract <E extends RealmModel> E copyOrUpdate(Realm realm, E object, boolean update, Map<RealmModel, RealmObjectProxy> cache);
 
+    public abstract void insertToRealm(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
```
Please add Javadoc for all of these

cmelchior replied:
insertObject?

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -104,6 +105,14 @@
+
+    public abstract void insertOrUpdateToRealm(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
+
+    public abstract void insertOrUpdateToRealm(Realm realm, RealmModel object, Iterator<? extends RealmModel> objects, Map<RealmModel, Long> cache);
```
insertCollection?

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+
+    /**
+     * Insert of a RealmObject, this is faster than {@link #copyToRealm(RealmModel)} since it
+     * doesn't return the inserted elements, and doesn't perform checks (threads, parameters etc.).
```
, and performs minimum allocations and checks.

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+    }
+
+    /**
+     * Insert of a RealmObject, this is faster than {@link #copyToRealm(RealmModel)} since it
```
Insert an unmanaged RealmObject,

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+     *
+     * @param object RealmObjects to insert.
+     */
+    public void insertToRealm(RealmModel object) {
```
`insertObject()` ?

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+     */
+    public void insertToRealm(RealmModel object) {
+        Map<RealmModel, Long> cache = new IdentityHashMap<RealmModel, Long>();
+        configuration.getSchemaMediator().insertToRealm(this, object, cache);
```
Make this a no-op if the input object is already managed.

nhachicha replied:
I intentionally disabled all checks for the sake of speed, if we start checking here we need to be consistent and perform checks in the generated code as well.  This is part of the tradeoff, if the user pass a `null` it's going to throw an NPE anyway later. Same argument for managed object.

If you want safety, use the pessimistic method `copyToRealm` that perform all kind of logical checks.

cmelchior replied:
I agree to a certain point. IMO it would not be acceptable if this method could SIGSEGV for example. I think we need unit test for most common error scenarios. `null`, managed objects (this will be a copy I guess), `null` values for Required objects. Possible more?

beeender replied:
Anything could cause a corrupted db if we don't do checks?

cmelchior replied:
No, it would just segfault during the transaction... the file itself should never get corrupted.

beeender replied:
Even copy from a RealmObject belongs to the other thread won't cause a corrupted db?

cmelchior replied:
No, you would call the setter on the RealmObject from the other thread which will throw an IllegalStateException

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+    }
+
+    /**
+     * Bulk insert or update of a list of RealmObjects, this is faster than {@link #copyToRealmOrUpdate(Iterable)} since it
```
of a collection of RealmObjects

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+
+    /**
+     * Bulk insert or update of a list of RealmObjects, this is faster than {@link #copyToRealmOrUpdate(Iterable)} since it
+     * doesn't return the inserted elements, and doesn't perform checks (threads, parameters etc.).
```
, and performs minimum allocations and checks.

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+     *
+     * @param objects RealmObjects to insert.
+     */
+    public void insertOrUpdateToRealm(Collection<? extends RealmModel> objects) {
```
insertOrUpdateObject()? / insertOrUpdate()? The prefix "toRealm" sounds really weird. I see the connection to `copyToRealmOrUpdate` but IMO I don't think there is any doubt there that `insert` refers to the Realm.

nhachicha replied:
I'm fine with dropping the `ToRealm`

Zhuinden replied:
Well considering `copyToRealmOrUpdate()` has Realm in it and so does `deleteFromRealm()` in each of its instances (`deleteAllFromRealm()`, `deleteFirstFromRealm()`), I think dropping `Realm` from it is odd from an API standpoint.

Please note I am not a Realm person as usual, but `insertIntoRealm()` makes most sense to me personally. Especially over just something generic, say `insert()`

spl replied:
Other options: insertRealmObject and insertRealmCollection

Zhuinden replied:
That makes sense too!

cmelchior replied:
IMO our API is already a bit to verbose in places. No need to go further down that route 😄 . `insertRealmObject(RealmModel obj)` and `insertRealmCollection(Collection<RealmModel> collection)` basically say the same thing twice (name + parameter).

If you look at the combine method description I am strongly leaning towards:
- `insert(RealmModel obj)`
- `insertOrUpdate(RealmModel obj)`
- `insert(Collection<RealmModel> collection)`
- `insertOrUpdate(Collection<RealmModel> collection)`

They are descriptive without being verbose and sets the apart from `copy<To/From>Realm` also has benefits I think. Also the name is similar to what you have in SQLite and other ORM's.

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+        Map<RealmModel, Long> cache = new IdentityHashMap<RealmModel, Long>(objects.size());
+        if (iterator.hasNext()) {
+            // access the first element to figure out the type
+            object = iterator.next();
```
IMO this check should be in the mediator instead, since that is what it's primary job is.

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+
+    /**
+     * Insert or update a RealmObject, this is faster than {@link #copyOrUpdate(RealmModel, boolean)}} since it
+     * doesn't return the inserted elements, and doesn't perform checks (threads, parameters etc.).
```
Same

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +785,82 @@ private Scanner getFullStringScanner(InputStream in) {
+     */
+    public void insertOrUpdateToRealm(RealmModel object) {
+        Map<RealmModel, Long> cache = new IdentityHashMap<RealmModel, Long>();
+        configuration.getSchemaMediator().insertOrUpdateToRealm(this, object, cache);
```
Check if object is already managed or null

nhachicha replied:
see my previous point

cmelchior commented:
Done with round 1. Generally looks really 💯 . Not sure we can optimize this much further, unless we start investigating special data structures for the cache, which is probably not worth it at this point.

Some general comments:
- We generally indent the JavaWriter code to follow the outline of the generated code. Makes it a lot easier to read.
- Naming of the public methods could use some discussion I think
- Remember to check for managed/null objects in the public API before calling the proxy classes

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,491 @@
+        realm.insertToRealm(allTypes);
+        realm.commitTransaction();
+
+        AllTypes realmTypes = realm.where(AllTypes.class).findFirst();
```
Perhaps use `AllJavaTypes` instead since right now there is some code paths in the annotation processor not getting touched (int/short)

dalinaum commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if ("java.lang.Integer".equals(fieldType)) {
+            writer
+                    .emitStatement("Integer %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
`"Integer %s =  ((%s)object).%s()"` -> `"Integer %s = ((%s)object).%s()"`

dalinaum commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if ("java.lang.Short".equals(fieldType)) {
+            writer
+                    .emitStatement("Short %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
Same above.

dalinaum commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if("java.lang.Double".equals(fieldType)) {
+            writer
+                    .emitStatement("Double %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
Same above

dalinaum commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -671,6 +674,517 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+
+        } else if ("java.util.Date".equals(fieldType)) {
+            writer
+                    .emitStatement("java.util.Date %s =  ((%s)object).%s()", getter, interfaceName, getter)
```
white space after `=`

kneth commented:
Is looking good to me.

nhachicha commented:
retest this please

nhachicha commented:
@realm/java updated

beeender commented:
I guess the CI fails because of some JNI function definition mismatch. Easy way to fix: just generate the header file with javah, and overwrite the original ones. Then build it. Compiler will give you an error about what is wrong.

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -846,10 +816,10 @@ private void emitInsertMethod(JavaWriter writer) throws IOException {
                         .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
-                        .beginControlFlow("for (%s %sItem : %sList)", Utils.getGenericType(field), fieldName, fieldName)
-                        .emitStatement("Long cacheItemIndex%s = cache.get(%sItem)", fieldName, fieldName)
+                        .beginControlFlow("for (%s %2$sItem : %2$sList)", Utils.getGenericType(field), fieldName)
```
First `%s` should be `%1$s` for readability IMO

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -913,10 +883,10 @@ private void emitInsertListMethod(JavaWriter writer) throws IOException {
                         .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
-                        .beginControlFlow("for (%s %sItem : %sList)", Utils.getGenericType(field), fieldName, fieldName)
-                        .emitStatement("Long cacheItemIndex%s = cache.get(%sItem)", fieldName, fieldName)
+                        .beginControlFlow("for (%s %2$sItem : %2$sList)", Utils.getGenericType(field), fieldName)
```
Same

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java:
```diff
@@ -116,15 +117,25 @@ public void insertToRealm(Realm realm, RealmModel type, Map<RealmModel, Long> ca
-            AllTypesRealmProxy.insert(realm, objects, cache);
-        } else {
-            throw getMissingProxyClassException(clazz);
+    public void insertToRealm(Realm realm, Collection<? extends RealmModel> objects) {
```
insert

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java:
```diff
@@ -116,15 +117,25 @@ public void insertToRealm(Realm realm, RealmModel type, Map<RealmModel, Long> ca
+        if (iterator.hasNext()) {
+            //  access the first element to figure out the type
+            object = iterator.next();
+            insertOrUpdateToRealm(realm, object, cache);
```
Isn't this a problem? It means that if you have primary keys it will update behind the scene instead of crashing because you tried to insert an object with a similar ID.

We had the same discussion for `copyToRealm` and `copyToRealmOrUpdate` and right now that doesn't automatically switch to copyOrUpdate, so IMO this should have the same semantics.

Personally I think we should revisit those for 2.x, but for now I think they should behave the same way.

nhachicha replied:
actually that's an error on the `RealmProxyMediatorGenerator` it should be `insert` good catch

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java:
```diff
@@ -116,15 +117,25 @@ public void insertToRealm(Realm realm, RealmModel type, Map<RealmModel, Long> ca
+        RealmModel object = null;
+        Map<RealmModel, Long> cache = new IdentityHashMap<RealmModel, Long>(objects.size());
+        if (iterator.hasNext()) {
+            //  access the first element to figure out the type
```
Not sure  I understand this. How will this help figuring out the type?

nhachicha replied:
so you can determine the `class`  here
```java
Class<RealmModel> clazz = (Class<RealmModel>) ((object instanceof RealmObjectProxy) ? object.getClass().getSuperclass() : object.getClass());
```

cmelchior replied:
Ok, but you don't use any return values from that method. Why not just folding it into the loop below?

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java:
```diff
@@ -142,15 +153,25 @@ public void insertOrUpdateToRealm(Realm realm, RealmModel obj, Map<RealmModel, L
-            AllTypesRealmProxy.insertOrUpdate(realm, objects, cache);
-        } else {
-            throw getMissingProxyClassException(clazz);
+    public void insertOrUpdateToRealm(Realm realm, Collection<? extends RealmModel> objects) {
```
insertOrUpdate

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * @param realm RealmObjects to insert.
+     * @param objects Collection of RealmObjects to insert or update.
+     */
+    public abstract void insertOrUpdateToRealm(Realm realm, Collection<? extends RealmModel> objects);
```
insertOrUpdate

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * @param realm RealmObjects to insert.
+     * @param objects Collection of RealmObjects to insert or update.
+     */
+    public abstract void insertToRealm(Realm realm, Collection<? extends RealmModel> objects);
```
insert

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * @param object RealmObjects to insert.
+     * @param cache the cache for mapping between unmanaged objects and their table row index for eventual reuse.
+     */
+    public abstract void insertToRealm(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
```
insert

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * @param object RealmObjects to insert.
+     * @param cache the cache for mapping between unmanaged objects and their table row index for eventual reuse.
+     */
+    public abstract void insertOrUpdateToRealm(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
```
insertOrUpdate

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +784,64 @@ private Scanner getFullStringScanner(InputStream in) {
+     * Insert or update an unmanaged RealmObject, this is generally faster than {@link #copyOrUpdate(RealmModel, boolean)}} since it
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     * <p>
+     * Please note, this doesn't check if the object is already managed or not.
```
Specifically, this doesn't check if the object is already managed, so inserting a managed object will duplicate it unlike `copyToRealm`.

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +784,64 @@ private Scanner getFullStringScanner(InputStream in) {
+    }
+
+    /**
+     * Bulk insert or update of a list of unmanaged RealmObjects, this is generally faster than {@link #copyToRealmOrUpdate(Iterable)} since it
```
Bulk insert or update of a list -> Insert or update a list of unmanaged

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +784,64 @@ private Scanner getFullStringScanner(InputStream in) {
+     * Bulk insert or update of a list of unmanaged RealmObjects, this is generally faster than {@link #copyToRealmOrUpdate(Iterable)} since it
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     * <p>
+     * Please note, this doesn't check if the object is already managed or not.
```
See below

cmelchior commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -784,6 +784,64 @@ private Scanner getFullStringScanner(InputStream in) {
     }
 
     /**
+     * Bulk insert of a list of unmanaged RealmObjects, this is generally faster than {@link #copyToRealm(Iterable)} since it
```
See below

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,584 @@
+    }
+
+    @Test
+    public void testManagedObject() {
```
Add a comment about expected outcome. e.g ``//Inserting a managed object will result in it being copied.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,584 @@
+        realm.beginTransaction();
+        try {
+            realm.insert(list);
+        } catch (NullPointerException ignore) {
```
add `fail()` after insert

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,584 @@
+    }
+
+    @Test
+    public void insertToRealm_CyclicType() {
```
Same test missing for `insertOrUpdate`

nhachicha commented:
@realm/java 
- Update as per feedback
- Fixed a bug in where Cyclic dependency graph could insert duplicate elements for both the new `insert` as well as the existing `copyToRealm` method (solved by sharing the cache Map between invocation)
- Updated the benchmark

dalinaum commented on CHANGELOG.md:
```diff
@@ -43,7 +44,7 @@ No changes since 0.91.1.
 ### Bug fixes
 
-* Fixed a bug when opening a Realm cause a staled memory mapping. Symptoms are error messages like "Bad or incompatible history type", "File format version doesn't match", and "Encrypted interprocess sharing is currently unsupported". 
+* Fixed a bug when opening a Realm cause a staled memory mapping. Symptoms are error messages like "Bad or incompatible history type", "File format version doesn't match", and "Encrypted interprocess sharing is currently unsupported".
```
`Fixed a bug when opening a Realm cause a staled memory mapping.`
`staled` -> `stalled`
`causes` seems more proper, but I am not good at English.

cmelchior replied:
Also add a entry about `copyToRealm` being fixed.

dalinaum commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ### Enhancements
 
+* Add `insert(RealmModel obj)`, `insertOrUpdate(RealmModel obj)`, `insert(Collection<RealmModel> collection)` and `insertOrUpdate(Collection<RealmModel> collection)` to perform batch inserts (#1684)
```
Add -> Added?

bmunkholm replied:
perform -> perform faster

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -1043,111 +1052,111 @@ private void emitInsertOrUpdateListMethod(JavaWriter writer) throws IOException
+                            .emitEmptyLine()
+                            .emitStatement("%s %sObj = ((%s) object).%s()", fieldType, fieldName, interfaceName, getter)
+                            .beginControlFlow("if (%sObj != null)", fieldName)
+                            .emitStatement("Long cache%1$s = cache.get(%1$sObj)", fieldName)
```
Please indent these so they follow the generated code.

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -1043,111 +1052,111 @@ private void emitInsertOrUpdateListMethod(JavaWriter writer) throws IOException
+                            .emitStatement("RealmList<%s> %sList = ((%s) object).%s()",
+                                    Utils.getGenericType(field), fieldName, interfaceName, getter)
+                            .beginControlFlow("if (%sList != null)", fieldName)
+                            .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
Same

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java:
```diff
@@ -104,7 +104,7 @@ public String getTableName(Class<? extends RealmModel> clazz) {
 
     @Override
-    public void insertToRealm(Realm realm, RealmModel type, Map<RealmModel, Long> cache) {
+    public void insert(Realm realm, RealmModel type, Map<RealmModel, Long> cache) {
```
Minor. I would probably rename `type` -> `obj`, because for the single object case it is also the object we actually insert.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1143,6 +1143,50 @@ public void copyToRealm_cyclicObjectReferences() {
         assertEquals("Two", realmObject.getObject().getName());
         assertEquals(2, realm.where(CyclicType.class).count());
+
+        realm.beginTransaction();
```
Either split this into a another unit test or add a comment about what is now being tested.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:
```diff
@@ -1143,6 +1143,50 @@ public void copyToRealm_cyclicObjectReferences() {
+        assertEquals("Two", realmObject.getObject().getName());
+        assertEquals(2, realm.where(CyclicTypePrimaryKey.class).count());
+
+        realm.beginTransaction();
```
Same

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .beginControlFlow("if (%sList != null)", fieldName)
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
+                        .beginControlFlow("for (%1$s %2$sItem : %2$sList)", Utils.getGenericType(field), fieldName)
+                        .emitStatement("Long cacheItemIndex%1$s = cache.get(%1$sItem)", fieldName)
```
Same for all of these

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("Float %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetFloat(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("Boolean %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetBoolean(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("byte[] %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetByteArray(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("java.util.Date %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetTimestamp(tableNativePtr, columnInfo.%sIndex, rowIndex, %s.getTime())", fieldName, getter);
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("String %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetString(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .emitEmptyLine()
+                        .emitStatement("%s %sObj = ((%s) object).%s()", fieldType, fieldName, interfaceName, getter)
+                        .beginControlFlow("if (%sObj != null)", fieldName)
+                        .emitStatement("Long cache%1$s = cache.get(%1$sObj)", fieldName)
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .emitEmptyLine()
+                        .emitStatement("%s %sObj = ((%s) object).%s()", fieldType, fieldName, interfaceName, getter)
+                        .beginControlFlow("if (%sObj != null)", fieldName)
+                        .emitStatement("Long cache%1$s = cache.get(%1$sObj)", fieldName)
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .emitStatement("RealmList<%s> %sList = ((%s) object).%s()",
+                                Utils.getGenericType(field), fieldName, interfaceName, getter)
+                        .beginControlFlow("if (%sList != null)", fieldName)
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
indent

cmelchior commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+                        .emitStatement("RealmList<%s> %sList = ((%s) object).%s()",
+                                Utils.getGenericType(field), fieldName, interfaceName, getter)
+                        .beginControlFlow("if (%sList != null)", fieldName)
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
indent

kneth commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -1043,111 +1052,111 @@ private void emitInsertOrUpdateListMethod(JavaWriter writer) throws IOException
-                            .nextControlFlow("else")
-                            .emitStatement("rowIndex = Table.nativeFindFirstString(tableNativePtr, pkColumnIndex, primaryKeyValue)")
-                            .endControlFlow();
+            if (hasPrimaryKey) {
```
Why indent here?

bmunkholm commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,67 +609,555 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        writer.emitStatement("object = (%s) objects.next()", className);
+        writer.beginControlFlow("if(!cache.containsKey(object))");
+
+            if (hasPrimaryKey) {
```
outdent section

bmunkholm commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -0,0 +1,616 @@
+        assertEquals(1, realm.where(AllTypesPrimaryKey.class).count());
+        AllTypesPrimaryKey obj = realm.where(AllTypesPrimaryKey.class).findFirst();
+
+        // Check that the the only element has all its properties updated
```
wording

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -774,16 +774,75 @@ private Scanner getFullStringScanner(InputStream in) {
+    }
+
+    /**
+     * Insert or update of a list of unmanaged RealmObjects. This is generally faster than {@link #copyToRealmOrUpdate(Iterable)} since it
```
"update of a" -> "update a"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -774,16 +774,75 @@ private Scanner getFullStringScanner(InputStream in) {
+    }
+
+    /**
+     * Insert of unmanaged RealmObject. This is generally faster than {@link #copyToRealm(RealmModel)} since it
```
of -> an

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
     public abstract <E extends RealmModel> E copyOrUpdate(Realm realm, E object, boolean update, Map<RealmModel, RealmObjectProxy> cache);
 
     /**
+     * Insert of an unmanaged RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
```
"Insert of an" -> "Insert an"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+    public abstract void insert(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
+
+    /**
+     * Insert or update a RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
```
", this" -> ". This"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * Insert of an unmanaged RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     *
+     * @param realm RealmObjects to insert.
```
description totally wrong.

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * Insert or update a RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     *
+     * @param realm RealmObjects to insert.
```
description totally wrong.

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     *
+     * @param realm RealmObjects to insert.
+     * @param object RealmObjects to insert.
```
RealmObjects -> RealmObject

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
+     *
+     * @param realm RealmObjects to insert.
+     * @param object RealmObjects to insert.
```
RealmObjects -> RealmObject. Or shouldn't it be "Realm object"?

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+    public abstract void insertOrUpdate(Realm realm, Collection<? extends RealmModel> objects);
+
+    /**
+     * Insert of a RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
```
-> "Insert a RealmObject. This"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -105,6 +106,44 @@
+    public abstract void insertOrUpdate(Realm realm, RealmModel object, Map<RealmModel, Long> cache);
+
+    /**
+     * Insert or update a RealmObject, this is generally faster than {@link #copyOrUpdate(Realm, RealmModel, boolean, Map)} since it
```
", this" -> ". This"

bmunkholm commented:
Reviewers: Please ensure to review the method documentation in details :-) Seems like a lot was missed.

bmunkholm commented:
Could we deprecate the copyToRealm()? If not, it's very unclear what the risk is when using the new faster inserts.

bmunkholm commented:
Was it considered to name the methods `add` instead of `insert`?

nhachicha commented:
@realm/java 
> Was it considered to name the methods add instead of insert?
consensus was on using `insert*` as it's similar to the insert query on SQL (no returned data)

> Could we deprecate the copyToRealm()? If not, it's very unclear what the risk is when using the new faster inserts.

`copyToRealm` return the managed RealmObject[s], it also perform all kinds of checks we don't do on `insert` as stated by @cmelchior you can't have safe & speed

Zhuinden commented:
Out of curiosity if this is the right place to ask,  is there a way to somehow add a managed Realm object's link to an inserted Realm object without having to find the object and setting it (for example with primary key)? 

Bulk insert would fix the issue of creating a new proxy for each inserted item,  but you would still need a proxy for every element you want to link the inserted object to.

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -785,7 +785,7 @@ private Scanner getFullStringScanner(InputStream in) {
 
     /**
-     * Insert of a list of unmanaged RealmObjects. This is generally faster than {@link #copyToRealm(Iterable)} since it
+     * Insert a list of an unmanaged RealmObjects. This is generally faster than {@link #copyToRealm(Iterable)} since it
```
remove "an"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -799,7 +799,7 @@ public void insert(Collection<? extends RealmModel> objects) {
 
     /**
-     * Insert of unmanaged RealmObject. This is generally faster than {@link #copyToRealm(RealmModel)} since it
+     * Insert of an unmanaged RealmObject. This is generally faster than {@link #copyToRealm(RealmModel)} since it
```
remove "of"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -814,7 +814,7 @@ public void insert(RealmModel object) {
 
     /**
-     * Insert or update of a list of unmanaged RealmObjects. This is generally faster than {@link #copyToRealmOrUpdate(Iterable)} since it
+     * Insert or update a list of an unmanaged RealmObjects. This is generally faster than {@link #copyToRealmOrUpdate(Iterable)} since it
```
remove "an"

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -106,39 +106,39 @@
      *
-     * @param realm RealmObjects to insert.
-     * @param object RealmObjects to insert.
+     * @param realm the reference to {@link Realm} where to create the object.
```
I can see the description is copied from the other methods, but it doesn't read well as a sentence.
How about: "@param realm reference to the {@link Realm} where the objects will be inserted.
Same goes for the other methods, but adjust singular/plural of "objects".

copyOrUpdate() is by the way also missing the description of this parameter.

bmunkholm commented on realm/realm-library/src/main/java/io/realm/internal/RealmProxyMediator.java:
```diff
@@ -106,39 +106,39 @@
-     * @param realm RealmObjects to insert.
-     * @param object RealmObjects to insert.
+     * @param realm the reference to {@link Realm} where to create the object.
+     * @param object RealmObject to insert.
```
Shouldn't it be a link? "@{link RealmObject}" instead? (same the other places)

bmunkholm commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,137 +609,631 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            writer
+                    .emitStatement("Double %s = ((%s)object).%s()", getter, interfaceName, getter)
+                    .beginControlFlow("if (%s != null)", getter)
+                    .emitStatement("Table.nativeSetDouble(tableNativePtr, columnInfo.%sIndex, rowIndex, %s)", fieldName, getter);
```
indent

bmunkholm commented:
@nhachicha We still need to update the docs to clearly describe what the difference is between the two ways of adding objects. What can go wrong (i.e. what do the programmer have to guarantee) when using these methods? What will happen if he/she does?

bmunkholm commented:
Re naming: Wouldn't it still be better to call it `add`? That would be consistent with our RealmList.

bmunkholm commented:
Love the speedups! Great work!

cmelchior commented:
Regarding naming there can be arguments for both `add` and `insert`.

- Following the Java Collection API's, the methods should be named `add(Object), addAll(Collection)`. `addOrUpdate` doesn't exists in the collection API. Neither `addAllOrUpdate()` or `addOrUpdateAll()` exists, nor is there a clear winner between them.

- Using overrides so we just had `add(Object), add(Collection), addOrUpdate(Object), addOrUpdate(Collection)`, would make it look similar to `copyToRealm()`, but not the standard Java Collection API.

- Swift uses `realm.add(obj, update: true)`. Objective C uses `addOrUpdateObject()`. OrmLite has `createOrUpdate()`. GreenDao has `insertOrReplace()` and `insertAndUpdate()`. Requery has `upsert()`.

- We moved away from the Java collection methods on Realm because we had conflicts with the interpretation, e.g. `clear()` became `delete()`. 

- The behaviour of calling `add` on a already managed object is perhaps a bit surprising as it would be duplicated instead of just being a no-op. Using `insert` makes that more explicit IMO.

- `insert` and `delete` are both concepts from SQL. `add` and `delete` is an odd mix.

- We stuck with `@PrimaryKey` instead of replacing with `@ObjectId`, i.e. staying with SQL terminology.

To summarize. There is precedence for using `add` from the other bindings, but no strong argument from already existing Realm Java API's. Given that people coming from SQL might be more familiar with INSERT, using `insertOrUpdate` might be easier to find.

So it is really a toss-up between the two and I don't have a strong preference either way, however I would probably lean towards `insertOrUpdate`

To be clear, I think the decision should be between:

1) `insert(Object), insert(Collection), insertOrUpdate(Object), insertOrUpdate(Collection)`
2) `add(Object), add(Collection), addOrUpdate(Object), addOrUpdate(Collection)`

Thoughts @realm/java @bmunkholm ?

spl commented:
As a new user of Realm, I appreciate the use of SQL terms that add the association of database query and update, reflecting the side effects these operations have. That is, I would expect to see `insert` more than `add`.

Zhuinden commented:
I personally think `insert` makes more sense.

kneth commented:
I don't have a strong opinion on this but lean mostly towards 1.

nhachicha commented:
@Zhuinden 
You can link an existing managed object to your unmanaged Realm object 
```java
        realm.beginTransaction();
        AllJavaTypes managedAllJavaTypes = realm.createObject(AllJavaTypes.class, 42);
        realm.commitTransaction();

        AllJavaTypes unmanagedObject = new AllJavaTypes(8);
        unmanagedObject.setFieldObject(managedAllJavaTypes);//Linking managed object to unmanaged object

        realm.beginTransaction();
        realm.insertOrUpdate(unmanagedObject);
        realm.commitTransaction();

        AllJavaTypes first = realm.where(AllJavaTypes.class).equalTo(AllJavaTypes.FIELD_LONG, 8).findFirst();
        assertNotNull(first);
        assertEquals(8, first.getFieldLong(), 0);
        assertNotNull(first.getFieldObject());
        assertEquals(42, first.getFieldObject().getFieldLong());
        assertEquals(2, realm.where(AllJavaTypes.class).count());
```
is this your use case? if not please comment with a unit test that describe your scenario 


@cmelchior +1 for `insert`

Zhuinden commented:
@nhachicha  essentially my usecase yes, but

        AllJavaTypes managedAllJavaTypes = realm.createObject(AllJavaTypes.class, 42);

which in my case would be obtained by 

      AllJavaTypes all = realm.where(AllJavaTypes.class).equalTo("id", 42).findFirst();

then you set it here 

       unmanagedObject.setFieldObject(managedAllJavaTypes);

But my question is whether something like this is possible in order to avoid the query:

       unmanagedObject.setFieldObject(AllJavaTypes.class, 42);

Because this way you avoid having to create the proxy for the object with the id 42

Of course, even with your version, I can reduce my current number of proxies from 4000 to about 150, so that's much better. I'm just curious at this point.

nhachicha commented:
@Zhuinden 
if you do the following you'll not create a proxy object.
```
AllJavaTypes  unmanagedLink = new AllJavaTypes(42);
unmanagedObject.setFieldObject(unmanagedLink);

realm.insertOrUpdate(unmanagedObject);
```
Actually in the entire implementation we don't create any proxy object, we directly create a new row (if necessary) and use the row index to update/insert the appropriate fields

Zhuinden commented:
@nhachicha Wait, so if you use `setFieldObject` with an unmanaged link object with a set id, then it won't replace the original `AllJavaTypes` in the database with an empty class (all fields cleared out), but instead use the ID field in `AllJavaTypes` to link the managed DB version of `AllJavaTypes` to `unmanagedObject`?

nhachicha commented:
@Zhuinden 
in this case it will replace with a cleared fields. The point was just to demonstrate how can you link within your object graph both managed & unmanaged objects.
of course, if you need to update a particular instance you'll need to find it first.

Zhuinden commented:
@nhachicha I see, that's a shame for my use-case (I wish I could just merge these objects!) but this is still an improvement on my side, thank you for the answers

nhachicha commented:
@realm/java

bmunkholm commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -538,7 +538,7 @@ public void testObjectNull() {
         try {
             realm.insert(nullObject);
             fail("Should trigger NullPointerException");
```
update text to "IllegalArgumentException" or remove it. It's not very useful to have anyway :-)

bmunkholm commented on realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java:
```diff
@@ -553,7 +553,7 @@ public void testListNull() {
         try {
             realm.insert(nullObjects);
             fail("Should trigger NullPointerException");
```
same as above

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -788,57 +788,101 @@ private Scanner getFullStringScanner(InputStream in) {
+     * <p>
+     * If you want the previous checks and the returned managed {@link RealmObject} use {@link #copyToRealm(Iterable)}, otherwise if
+     * you have a large number of object this method is generally faster.
      *
```
Shouldn't we have @see reference to copyToRealm.
And similar a reference in `copyToRealm` to this method.

bmunkholm commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -788,57 +788,101 @@ private Scanner getFullStringScanner(InputStream in) {
      * @param objects RealmObjects to insert.
      */
     public void insert(Collection<? extends RealmModel> objects) {
+        checkIfValid();
```
I guess the argument for having insert() for one object is that you use this in a loop perhaps? If you do that how much faster is this method now? Maybe it's sufficiently valuable that it creates less objects to GC?
Is the one object variants at all relevant to have still? Or wouldn't the collection variant be enough?

Zhuinden replied:
Isn't `checkIfValid` the call that slowed down `copyToRealmOrUpdate` so much that a new `insert` method had to be added in the first place?

cmelchior replied:
It was part of it. Allocations were worse though, but this is just 1 call, it is not done recursively, so it will not have a measurable impact.

cmelchior replied:
@bmunkholm Single object insert is still useful of you have very deep object hierarchies, e.g. I have seen a fair number of REST API's that have a top-level wrapper object and then a lot of dangling objects under it. It also makes it consistent with copyToRealm

Zhuinden replied:
I also have a use-case for single object insert - and that is that having a list of all 4000 objects I'm storing in a batch is too much, so I'm streaming it directly one by one and persisting them to the Realm like so

cmelchior commented:
Except BM's comments about unit test and `@see` this looks 👍

bmunkholm commented:
Re naming, so far we have tried to have consistent naming across bindings, unless there were strong reason to not have that. That would move the needle to `add`. And keep it consistent with our own RealmList API.

Zhuinden commented:
@bmunkholm  the difference between `add` and `copyOrUpdate` isn't very clear :/

kneth commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ### Enhancements
 
+* Added `insert(RealmModel obj)`, `insertOrUpdate(RealmModel obj)`, `insert(Collection<RealmModel> collection)` and `insertOrUpdate(Collection<RealmModel> collection)` to perform batch inserts (#1684)
```
`.` is missing at the end

kneth commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -773,16 +775,123 @@ private Scanner getFullStringScanner(InputStream in) {
 
     /**
+     * Insert a list of an unmanaged RealmObjects. This is generally faster than {@link #copyToRealm(Iterable)} since it
+     * doesn't return the inserted elements, and performs minimum allocations and checks.
```
"performs" -> "perform"?

nhachicha replied:
_the method performs_  I think

kneth commented on realm/realm-library/src/main/java/io/realm/Realm.java:
```diff
@@ -773,16 +775,123 @@ private Scanner getFullStringScanner(InputStream in) {
+     * <li>Copying an object will copy all field values. Any unset field in the object and child objects will be set to their default value if not provided</li>
+     * </ul>
+     * <p>
+     * If you want the previous checks and the returned managed {@link RealmObject} use {@link #copyToRealm(Iterable)}, otherwise if
```
remove "previous"
"the returned managed {@link RealmObject}" -> "the managed {@link RealmObject} returned, "

kneth commented:
Besides my small comments, I think it is ready to me merged.

zaki50 commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,137 +609,631 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+        writer.emitStatement("Table table = realm.getTable(%s.class)", className);
+        writer.emitStatement("long tableNativePtr = table.getNativeTablePointer()");
+        writer.emitStatement("%s columnInfo = (%s) realm.schema.getColumnInfo(%s.class)"
+                , columnInfoClassName(), columnInfoClassName(), className);
```
first `,` should be at the end of previous line.

zaki50 commented:
Great work! :+1:

bmunkholm commented:
Re naming we decided to keep `ìnsert`.

zaki50 commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,137 +609,631 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
+            } else if (Utils.isRealmList(field)) {
+                writer
+                        .emitEmptyLine()
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
If a model class has multiple RealmList fields, a local variable `nativeLinkViewPtr`will be declared multiple time in the same scope.

zaki50 commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -606,137 +609,631 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
-                    .beginControlFlow("if (%sList != null)", fieldName)
-                        .emitStatement("RealmList<%s> %sRealmList = ((%s) realmObject).%s()",
+                        .emitEmptyLine()
+                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
```
If a model class has multiple RealmList fields, a local variable nativeLinkViewPtrwill be declared multiple time in the same scope.

nhachicha commented:
@zaki50 fixed

zaki50 commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -1006,8 +1006,8 @@ private void emitInsertOrUpdateMethod(JavaWriter writer) throws IOException {
                         .emitEmptyLine()
-                        .emitStatement("long nativeLinkViewPtr = Table.nativeGetLinkView(tableNativePtr, columnInfo.%sIndex, rowIndex)", fieldName)
-                        .emitStatement("LinkView.nativeClear(nativeLinkViewPtr)")
+                        .emitStatement("long nativeLinkViewPtr%1$s = Table.nativeGetLinkView(tableNativePtr, columnInfo.%1$sIndex, rowIndex)", fieldName)
```
How about `%1$sNativeLinkViewPtr` instead of `nativeLinkViewPtr%1$s`.
`%1$sNativeLinkViewPtr` is more natural as a variable name `columnRealmListNativeLinkViewPtr` vs `nativeLinkViewPtrcolumnRealmList`.

zaki50 commented:
Other than my last comment (https://github.com/realm/realm-java/pull/2999#discussion_r69055151),
I vote :+1:
